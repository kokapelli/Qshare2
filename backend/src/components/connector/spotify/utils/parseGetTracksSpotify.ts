import { QShareTrack, QShareArtist } from "../../../music/musicTypes";

interface SpotifyTrack {
    name: string;
    uri: string;
    artists: SpotifyArtist[];
    album: SpotifyAlbum;
    popularity: number;
}

interface SpotifyArtist {
    uri: string;
    name: string;
}

interface SpotifyAlbum {
    images: SpotifyAlbumImage[];
}

interface SpotifyAlbumImage {
    width: number;
    height: number;
    url: string;
}

interface SpotifyGetTracksResponse {
    tracks: {
        name: string
        items: SpotifyTrack[]
        limit: number
    };
}

export interface QShareGetTrackResponse extends QShareTrack {
    popularity: number;
}

export interface QShareGetTracksResponse {
    limit: number;
    tracks: QShareGetTrackResponse[];
}

export const parseGetTracksSpotify = (data: SpotifyGetTracksResponse) => {
    const response: QShareGetTracksResponse = {} as any;
    const tracks: QShareGetTrackResponse[] = [];

    response.tracks = tracks;
    response.limit = data.tracks.limit;
    for (const track of Array.from(data.tracks.items)) {
        const newTrack: QShareGetTrackResponse = {} as any;
        newTrack.name = track.name;

        const newArtists: QShareArtist[] = [];
        for (const artist of track.artists) {
            const newArtist: SpotifyArtist = {
                uri: artist.uri,
                name: artist.name
            };

            newArtists.push(newArtist);
        }

        newTrack.artists = newArtists;

        const newImages = [];
        for (const image of track.album.images) {
            const newImage: SpotifyAlbumImage = {
                height: image.height,
                width: image.width,
                url: image.url
            };
            newImages.push(newImage);
        }

        newTrack.images = newImages;

        newTrack.uri = track.uri;
        newTrack.popularity = track.popularity;
        tracks.push(newTrack);
    }

    return response;
};
