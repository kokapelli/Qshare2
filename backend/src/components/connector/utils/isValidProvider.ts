import { SPOTIFY_PROVIDER } from "../../constants/providers";

const isValidProvider = (provider: string) => {
    return provider === SPOTIFY_PROVIDER;
};

export { isValidProvider };