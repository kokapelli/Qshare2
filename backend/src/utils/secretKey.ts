import crypto from "crypto";

const SECRET_KEY = crypto.randomBytes(256);

export default SECRET_KEY;
