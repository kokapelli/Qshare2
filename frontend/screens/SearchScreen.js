import React, { Component } from 'react'
import { StyleSheet, View, Text, TextInput, FlatList, TouchableOpacity, Alert } from 'react-native'
import { Icon } from 'expo'
import SongButton from '../components/SearchSongButton'
import { connect } from 'react-redux'
import { setSearchQuery, searchSongs } from '../actions/search'
import {addSong} from '../actions/queue'

/*Insert filter connected to Spotify API */
export class SearchScreen extends Component {
  static navigationOptions = {
    header: null,
  }

  componentWillReceiveProps (newProps){
    const {songAdded:songAddedOld, errorAdding:errorAddingOld} = this.props;
    const {songAdded,errorAdding} = newProps;

    //if song added songAddedOld = false, songAdded = true
    if (!songAddedOld && songAdded && !errorAdding){
      //display alert that it was successful.
      Alert.alert(
        'Queue',
        'Song added to queue.',
        [],
        {cancelable: true},
      );
    }

    if (!songAddedOld && !songAdded && errorAdding) {
      Alert.alert(
        'Queue',
        'Oh no! Something went wrong!',
        [],
        {cancelable: false},
      );
    }
  }

  onSubmit () {
    const {query} = this.props
    this.props.searchSongs(query)
  }

  _keyExtractor = (item, index) => {
    return item.uri
  }

  _renderItem = ({item}) => {
    return <SongButton item={item}/>
  }

  render () {
    const {query, songs} = this.props
    return (
      <View style={styles.container}>
        <View style={styles.searchContainer}>
          <View style={styles.buttonContainer}>
            <TouchableOpacity testID = 'searchButton' onPress={() => {}}>
              <Icon.Ionicons name={'md-search'} size={35} color="#999999"/>
            </TouchableOpacity>
          </View>
          <TextInput
            testID='searchInput' 
            style={styles.searchInput}
            placeholder="Search for a song"
            placeholderTextColor="gray"
            value={query}
            onChangeText={(text) => this.props.setSearchQuery(text)}
            onSubmitEditing={this.onSubmit.bind(this)}
          />
        </View>
        {
          songs && songs.length > 0 ? <View style={styles.headerContainer}>
            <Text testID='songsFound' style={styles.headerText}>Songs Found</Text>
          </View> : <Text/>
        }
        <View style={styles.resultsContainer}>
          <FlatList
            testID='songList' 
            data={songs}
            keyExtractor={this._keyExtractor}
            renderItem={this._renderItem}
          />
        </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#181818',
  },
  searchContainer: {
    flex: 1,
    flexDirection: 'row',
    borderColor: '#999999',
    borderWidth: 1,
    paddingLeft: 10,
    paddingRight: 10,
    borderRadius: 50,
    margin: 10,
  },
  searchInput: {
    flex: 5,
    marginTop: 5,
    fontSize: 25,
    color: '#ffffff',
    marginBottom: 5
  },
  buttonContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  headerContainer: {
    flex: 2,
    justifyContent: 'center',
  },
  headerText: {
    marginLeft: 15,
    fontSize: 35,
    color: 'white',
  },
  resultsContainer: {
    flex: 11,
    color: '#282828',
    borderColor: '#383838',
    flexDirection: 'column',
  },
})

const mapStateToProps = (state) => ({
  songs: state.search.songs,
  query: state.search.query,
  songAdded: state.queue.songAdded,
  errorAdding: state.queue.errorAdding
})

export default connect(mapStateToProps, {setSearchQuery, searchSongs, addSong})(SearchScreen)