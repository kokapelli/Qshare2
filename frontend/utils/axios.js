import Axios from "axios"
import { settings } from "../settings/settings";

const axios = Axios.create()
axios.defaults.headers.common['Provider'] = settings.provider

export default axios