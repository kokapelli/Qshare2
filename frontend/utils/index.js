export const findByTestAttr = (component, attr)  => {
    const wrapper = component.find(`[testID='${attr}']`);
    return wrapper
  }
  