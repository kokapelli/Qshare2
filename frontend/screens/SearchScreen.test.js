import { SearchScreen } from './SearchScreen'
import { shallow, configure } from 'enzyme'
import * as renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'
import { findByTestAttr } from '../utils/'

const initialState = {
  query: '',
  songs: []
}

const mockSongs = [{
      track: { name:"Test1" }
    },
    { track: { name:"Test2" }
    },
    { track: { name:"Test3" }
  }
] 

configure({ adapter: new Adapter() });

describe('SearchScreen Snapshot', () => {

  it('matches the snapshot', () => {
    const tree = renderer.create(
    <SearchScreen/>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('SearchScreen Units', () => {
  
  beforeEach(()=>{
    this.component = shallow(<SearchScreen props={initialState}/>)
  })

  it('Component renders properly', () => {
    const rendered = renderer.create(<SearchScreen/>).toJSON()
    expect(rendered).toBeTruthy()
  });

  it('Search button should be clickable', () => {
    const touchable = findByTestAttr(this.component, 'searchButton')
    expect(touchable.exists()).toEqual(true) 
    expect(touchable.simulate('click')).toBeDefined()
    expect(touchable.simulate('click')).toEqual({})
  })

  it('Search text field should contain placeholder', () => {
    const expected = "Search for a song"
    const textField = findByTestAttr(this.component, 'searchInput')
    expect(textField.props().placeholder).toEqual(expected)
  })

  it('Search text field should show query value', () => {
    const newQuery = "Test"
    this.component.setProps({ query: newQuery})
    const textField = findByTestAttr(this.component, 'searchInput')
    expect(textField.props().value).toEqual(newQuery)
  })

  it('Search text field query should change on query change', () => {
    const newQuery = "Test"
    let textField = findByTestAttr(this.component, 'searchInput')
    expect(textField.props().value).toBeUndefined()
    this.component.setProps({ ...initialState, query: newQuery})
    textField = findByTestAttr(this.component, 'searchInput')
    expect(textField.props().value).toEqual(newQuery)
  })
  
  it('Should print Songs Found when songs exist in the properties', () => {
    const expected = "Songs Found"
    this.component.setProps({ songs: mockSongs})
    const songsFound = findByTestAttr(this.component, 'songsFound')
    expect(songsFound.props().children).toEqual(expected)
  })

  it('Should display the correct amount of songs', () => {
    const expected = mockSongs.length
    this.component.setProps({ songs: mockSongs})
    const songList = findByTestAttr(this.component, 'songList')
    expect(songList.props().data).toHaveLength(expected)
  })

  it('Should display the correct songs', () => {
    this.component.setProps({ songs: mockSongs})
    const songList = findByTestAttr(this.component, 'songList')
    expect(songList.props().data[0].track.name).toEqual(mockSongs[0].track.name)
    expect(songList.props().data[1].track.name).toEqual(mockSongs[1].track.name)
    expect(songList.props().data[2].track.name).toEqual(mockSongs[2].track.name)
  })
})