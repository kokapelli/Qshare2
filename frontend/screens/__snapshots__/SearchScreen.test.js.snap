// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen Snapshot matches the snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#181818",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#999999",
        "borderRadius": 50,
        "borderWidth": 1,
        "flex": 1,
        "flexDirection": "row",
        "margin": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="searchButton"
      >
        <Text />
      </View>
    </View>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Search for a song"
      placeholderTextColor="gray"
      style={
        Object {
          "color": "#ffffff",
          "flex": 5,
          "fontSize": 25,
          "marginBottom": 5,
          "marginTop": 5,
        }
      }
      testID="searchInput"
      underlineColorAndroid="transparent"
    />
  </View>
  <Text />
  <View
    style={
      Object {
        "borderColor": "#383838",
        "color": "#282828",
        "flex": 11,
        "flexDirection": "column",
      }
    }
  />
</View>
`;
